curvefs/src/space/space_manager.h

```diff
@@ -0,0 +1,104 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef CURVEFS_SRC_SPACE_SPACE_MANAGER_H_
+#define CURVEFS_SRC_SPACE_SPACE_MANAGER_H_
+
+#include <bthread/mutex.h>
+
+#include <memory>
+#include <unordered_map>
+
+#include "curvefs/proto/mds.pb.h"
+#include "curvefs/src/space/bitmap_allocator.h"
+#include "curvefs/src/space/s3_allocator.h"
+#include "src/common/concurrent/rw_lock.h"
+
+namespace curvefs {
+namespace space {
+
+struct SpaceManagerOption {};
+
+class SpaceManager {
+ public:
+    SpaceManager() = default;
+    virtual ~SpaceManager() = default;
+
+    virtual SpaceStatusCode InitSpace(const mds::FsInfo& fsInfo) = 0;
+
+    virtual SpaceStatusCode UnInitSpace(uint32_t fsId, mds::FSType fsType) = 0;
+
+    virtual SpaceStatusCode StatSpace(uint32_t fsId, uint64_t* total,
+                                      uint64_t* available,
+                                      uint64_t* blkSize) = 0;
+
+    virtual SpaceStatusCode AllocateSpace(uint32_t fsId, uint64_t size,
+                                          const SpaceAllocateHint& hint,
+                                          Extents* exts) = 0;
+
+    virtual SpaceStatusCode DeallocateSpace(uint32_t fsId,
+                                            const ProtoExtents& extents) = 0;
+
+    virtual SpaceStatusCode AllocateS3Chunk(uint32_t fsId, uint64_t* id) = 0;
+
+ private:
+    SpaceManager(const SpaceManager&);
+    SpaceManager& operator=(const SpaceManager&);
+};
+
+class SpaceManagerImpl : public SpaceManager {
+ public:
+    explicit SpaceManagerImpl(const SpaceManagerOption& opt);
+
+    SpaceStatusCode InitSpace(const mds::FsInfo& fsInfo) override;
+
+    SpaceStatusCode UnInitSpace(uint32_t fsId, mds::FSType fsType) override;
+
+    SpaceStatusCode StatSpace(uint32_t fsId, uint64_t* total,
+                              uint64_t* available, uint64_t* blkSize) override;
+
+    SpaceStatusCode AllocateSpace(uint32_t fsId, uint64_t size,
+                                  const SpaceAllocateHint& hint,
+                                  Extents* exts) override;
+
+    SpaceStatusCode DeallocateSpace(uint32_t fsId,
+                                    const ProtoExtents& extents) override;
+
+    SpaceStatusCode AllocateS3Chunk(uint32_t fsId, uint64_t* id) override;
+
+ private:
+    SpaceStatusCode InitS3Space(const mds::FsInfo& fsInfo);
+
+    SpaceStatusCode InitBlockSpace(const mds::FsInfo& fsInfo);
+
+    SpaceStatusCode UnInitS3Space(uint32_t fsId);
+
+    SpaceStatusCode UnInitBlockSpace(uint32_t fsId);
+
+ private:
+    SpaceManagerOption opt_;
+
+    ::curve::common::BthreadRWLock blkRwlock_;
+    std::unordered_map<int, std::unique_ptr<BitmapAllocator>> blkAllocators_;
```

Comment:
The "fsId" is uint32_t, it is better to use uint32_t as the key in this map.

Reply:
fix